Managed RabbitMQ

bdero opened the pull request:
Implements support for a centrally managed RabbitMQ instance when not using the "ephemeral databases" option.

This PR relies on/uses a bunch of other PRs [linked from the ticket](https://tasks.opencraft.com/browse/OC-1719). Also, this code depends on changes made in #145.

**Testing instructions**:

1. Checkout this branch.
1. Add the following to your .env file:
   
   ```
   RABBITMQ_API_URL='https://rabbitmq-dev.opencraft.hosting:15671'
   RABBITMQ_ADMIN_USERNAME='[username]'
   RABBITMQ_ADMIN_PASSWORD='[password]'
   INSTANCE_RABBITMQ_HOST='rabbitmq-dev.opencraft.hosting'
   INSTANCE_RABBITMQ_PORT='5671'
   ```

1. Run `make migrate`.

1. Create a new instance and spawn an appserver. Set the following settings:

   ```
   Configuration source repo url: https://github.com/open-craft/configuration.git
   Configuration version: bdero/opencraft-rabbitmq-vhost-config

   Edx platform repository url: https://github.com/open-craft/edx-platform.git
   Edx platform commit: bdero/opencraft-celery-broker-tls
   ```

   **The following 4 steps can be done before the appserver finishes provisioning.**

1. Go to the [OpenEdXInstance admin panel](http://localhost:5000/admin/instance/openedxinstance/) and check if the vhost and user data was properly generated - this can be verified a couple of minutes after kicking off the appserver spawn task:

   ![screenshot](http://i.imgur.com/XsuYMZ5.png)

1. Navigate to https://rabbitmq-dev.opencraft.hosting:15671 in your browser and login to the administration account. The administration account credentials can be found by decrypting the [rabbitmq-dev host vars](https://github.com/open-craft/deployment-secrets-services/pull/13/files#diff-0b5b5280556d5a19a034420a0c90b5feR1).

1. Select the "Admin" tab and verify that the two new users were added to the new vhost.

1. Now, go grab some coffee/tea/beer. It's time to wait for the appserver to finish provisioning.

1. Once the appserver finishes provisioning, login to studio on the new Open edX instance and create a rerun of the default course.

1. Return to the RabbitMQ admin panel and verify that there was activity on the instance's new vhost.

**Reviewers**
- [ ] @haikuginger

haikuginger commented on instance/migrations/0063_auto_20161104_1720.py:
```diff
@@ -0,0 +1,54 @@
+            name='rabbitmq_provisioned',
+            field=models.BooleanField(default=False),
+        ),
+        migrations.AddField(
```
Might it be worth adding a Python migration that sets default values to refer to the local Rabbit instance?

bdero replied:
Hmm, I don't think so - we're going to roll this out to all instances immediately that are not using the "ephemeral databases" option, and I don't think there's a use case for keeping any non-ephemeral instances using the local RabbitMQ.

Also, these models are used for adding vhosts/users to the RabbitMQ setup configured in OCIM, so we'd need to have some extra checks/another boolean field in here if we wanted to support using local RabbitMQ setups on non-ephemeral instances.

haikuginger replied:
Sounds good!

haikuginger commented on instance/models/mixins/rabbitmq.py:
```diff
@@ -0,0 +1,105 @@
+    class Meta:
+        abstract = True
+
+    def _get_rabbitmq_vars(self, remove):
```
I'd rather have either a "more pure" Ansible implementation (less manual rendering in Python; render vars in a template via the Django templating system) or a more pure Python implementation (using the RabbitMQ HTTP API) - there are a few API implementations, or, if you want, it's simple enough to implement with Requests.

bdero replied:
Yeah I thought I might get a comment on this bit. ;) I'm inclined to agree. I'll opt for rendering the vars in a template - this will make it more legible and more in line with Ansible vars elsewhere in the codebase.

haikuginger replied:
That sounds good to me! Using the HTTP API would also be perfectly valid, similar to how we provision new users/databases for MySQL and Mongo.

antoviaque replied:
I would vote for consistency with the way we did this with other services - if the HTTP API allows to create user, it seems a bit overkill to use ansible for this?

bdero replied:
@antoviaque @haikuginger Alright, I replaced the playbook with calls the the RabbitMQ HTTP API (supplied by the management plugin). The API is accessible from here (HTTPS only): https://rabbitmq-dev.opencraft.hosting:15671/api/

I added some letsencrypt tasks and a RabbitMQ config template to get TLS working: https://github.com/open-craft/ansible-rabbitmq/pull/1/files

haikuginger commented on instance/models/openedx_instance.py:
```diff
@@ -331,6 +331,7 @@ def spawn_appserver(self):
             self.load_balancing_server = LoadBalancingServer.objects.select_random()
             self.save()
             self.load_balancing_server.reconfigure()
+            self.load_balancing_server.reconfigure()
```
Why twice?

haikuginger replied:
👍

haikuginger replied:
Still doubled up here. ;)

haikuginger commented on instance/models/rabbitmq.py:
```diff
@@ -0,0 +1,36 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Definition of RabbitQM-specific models.
```
s/RabbitQM/RabbitMQ/

haikuginger commented on instance/models/rabbitmq.py:
```diff
@@ -0,0 +1,36 @@
+    """
+    A model representing a single RabbitMQ user
+    """
+    username = models.CharField(max_length=32, unique=True)
```
We're enforcing username uniqueness, which means an IntegrityError will be raised if we get a duplicate. It's not likely to happen, but are we handling such an error and generating a different name if necessary? Alternatively, can we check to see if a user with a particular username exists before going to create a new one? (abstract the creation into a method that performs the check)

bdero replied:
Thanks, I added code to have it retry until a unique user is found.

bdero commented on instance/models/mixins/openedx_database.py:
```diff
@@ -234,6 +236,19 @@ def set_field_defaults(self):
             self.mongo_user = get_random_string(length=16, allowed_chars=string.ascii_lowercase)
             self.mongo_pass = get_random_string(length=32)
+        if not self.rabbitmq_vhost:
+            def new_rabbitmq_user():
```
This function is a replacement for other inline alternatives, where I would otherwise be looping over the two member fields I'm assigning (difficult to read), or cut/pasting the RabbitMQUser creation block (redundant and difficult to read).

I placed the function inline rather than as a method of the DatabaseMixin or elsewhere because I wanted to be clear that this function has no purpose anywhere else except for with the following two lines.

I reordered the placement of the fields to make this more clear. Do you think this makes sense? I'm also open to alternative patterns for creating these two users.

bdero replied:
This method became more complicated, so I will make this a static method of `OpenEdXDatabaseMixin`.

haikuginger replied:
Static method sounds good. Since you're subclassing ValidateModelMixin for the RabbitMQUser object, you could also generate the username and password on save with `set_field_defaults`.

haikuginger commented on instance/models/mixins/database.py:
```diff
@@ -231,3 +234,26 @@ def deprovision_mongo(self):
+    def deprovision_rabbitmq(self):
+        """
+        Delete RabbitMQ vhost and users
+        """
```
Newline at end of file, please.

bdero replied:
Sorry, I actually pushed an old version of the code, but it's fixed now. These functions are filled with something and the file has a newline.

haikuginger commented on instance/models/mixins/rabbitmq.py:
```diff
@@ -0,0 +1,105 @@
+    rabbitmq_vhost = models.CharField(max_length=16, blank=True)
+    rabbitmq_provider_user = models.ForeignKey(RabbitMQUser, related_name='provider_instance',
+                                               on_delete=models.CASCADE, blank=True, null=True)
+    rabbitmq_consumer_user = models.ForeignKey(RabbitMQUser, related_name='consumer_instance',
```
Nit: Please either fit all arguments on a single line, or, if there are too many, use a single line per argument.

bdero replied:
Done.

haikuginger commented on instance/models/mixins/openedx_database.py:
```diff
@@ -214,9 +217,26 @@ def generate_var_name(var):
+        Return a new RabbitMQ user.
+        """
+        user = None
+        while user is None:
```
I hate to harp on it, but we need to consider the degenerate case - let's limit the number of creation attempts, even though it realistically shouldn't make a difference.

bdero replied:
Done.

haikuginger commented:
Okay; I think I'm mostly happy with the code; I'll go ahead and test once you've updated the RabbitMQ provisioning stuff and I have a reliable Internet connection again (probably this evening).

antoviaque commented on instance/models/mixins/openedx_database.py:
```diff
@@ -24,15 +24,18 @@
 from django.utils.crypto import get_random_string
 
 from .database import MySQLInstanceMixin, MongoDBInstanceMixin
+from .rabbitmq import RabbitMQInstanceMixin
```
@bdero Can you use absolute imports? I know we still have a few relative ones (including the one right above) so you might have done this for consistency, but it would get us closer to be consistent to change the line above tp absolute, insteaad.

bdero replied:
Done.

antoviaque commented on opencraft/settings.py:
```diff
@@ -449,6 +449,12 @@
 INSTANCE_MONGO_URL_OBJ = urlparse(INSTANCE_MONGO_URL) if INSTANCE_MONGO_URL else None
 
+# Configure external RabbitMQ host
+RABBITMQ_ENABLE = env.bool('RABBITMQ_ENABLE', default=True)
```
What would be the cases where we would want to disable this?

bdero replied:
Probably nothing, so I just removed this.

antoviaque commented on opencraft/settings.py:
```diff
@@ -449,6 +449,12 @@
+RABBITMQ_ENABLE = env.bool('RABBITMQ_ENABLE', default=True)
+# The RabbitMQ host must be accessible from both OpenCraft IM as well as well as any instances using it.
+INSTANCE_RABBITMQ_HOST = env('INSTANCE_RABBITMQ_HOST', default=None)
+INSTANCE_RABBITMQ_SSH_USERNAME = env('INSTANCE_RABBITMQ_SSH_USERNAME', default='ubuntu')
```
Remember to add documentation to the README about these config variables.

antoviaque commented on instance/models/mixins/rabbitmq.py:
```diff
@@ -0,0 +1,115 @@
+        """
+        if settings.RABBITMQ_ENABLE and not self.rabbitmq_provisioned:
+            self._run_rabbitmq_playbook(False)
+        self.rabbitmq_provisioned = True
```
If `RABBITMQ_ENABLE` is False, we mark rabbitmq as provisioned even though we don't use it and haven't actually provisioned it?

bdero replied:
This setting is gone now.

antoviaque commented on opencraft/settings.py:
```diff
@@ -447,6 +453,10 @@
 
+# The RabbitMQ host must be accessible from both OpenCraft IM as well as well as any instances using it.
+INSTANCE_RABBITMQ_HOST = env('INSTANCE_RABBITMQ_HOST', default=None)
+INSTANCE_RABBITMQ_PORT = env('INSTANCE_RABBITMQ_PORT', default=5671)
```
The new variables still need doc - cf https://github.com/open-craft/opencraft/pull/152#discussion_r87029553

haikuginger commented on instance/models/mixins/rabbitmq.py:
```diff
@@ -0,0 +1,121 @@
+                self,
+                url,
+                action,
+                response.status
```
`.status_code`

haikuginger commented:
@bdero, since we've updated this to _not_ use the Vagrant provisioning method for Rabbit, I need the admin username and password for the existing setup. Can you either encrypt those with my public key and email them or add them to our KeePass database?

bdero commented:
@haikuginger The admin credentials are in the [instance host vars](https://github.com/open-craft/deployment-secrets-services/pull/13/files#diff-0b5b5280556d5a19a034420a0c90b5feR1) and the `vault-pass` is in the keepass. You can use use `ansible-vault view --vault-password-file=$VAULT_PASS` to view it with the vault password in place.

haikuginger commented:
Okay, I was able to test this and verified that the new vhost was created. My connection seems to be having issues, the symptoms of which are that I'm unable to fully provision a new instance from my local OCIM instance, but things do appear to be getting set up properly when looking at the variables, with one exception - `XQUEUE_RABBITMQ_PORT` is being set to the hostname, which seems undesirable.

I'd also like to see testing to ensure the provisioning process - for an example, take a look at the tests for MySQL and MongoDB, and of course, we need to pass the existing tests.

bdero commented:
@haikuginger I addressed all of the comments: Existing tests are fixed, I added new tests, documented stuff in the readme, and we should have a passing build now.

I also updated the description of this PR to contain working build steps. I've made a bunch of minor changes to the other PRs floating around as well. I've been trying continuously for many hours to get a single completed successful build in a local setup where I can actually test these changes all the way though again, but between external service issues and OVH sometimes refusing to launch instances, I haven't been able to get a new working build such that I can test instance connectivity with rabbit.

I was able to identify the issues you ran into and supply fixes, and I'm hopeful that the build I'm currently running will finally complete. In the meantime, this is fully ready for review. Any issues that we run into on running instances while trying to test should be minor configuration problems at worst.

haikuginger commented on instance/models/mixins/openedx_database.py:
```diff
@@ -285,4 +314,15 @@ def get_database_settings(self):
+        # RabbitMQ:
+        template = loader.get_template('instance/ansible/rabbitmq.yml')
+        new_settings += template.render({
+            'vhost': self.rabbitmq_vhost,
```
Need to add the instance port to the context object.

haikuginger commented:
@bdero, I was able to provision an instance, but got the following stack trace when I tried to set up a rerun of a course:

```python
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic-2.64.0.48/newrelic/hooks/framework_django.py", line 503, in wrapper
    return wrapped(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 22, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/course.py", line 262, in course_handler
    return _create_or_rerun_course(request)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/util/json_request.py", line 52, in parse_json_into_request
    return view_function(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/course.py", line 752, in _create_or_rerun_course
    return _rerun_course(request, org, course, run, fields)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/course.py", line 861, in _rerun_course
    rerun_course.delay(unicode(source_course_key), unicode(destination_course_key), request.user.id, json_fields)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py", line 453, in delay
    return self.apply_async(args, kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py", line 559, in apply_async
    **dict(self._get_exec_options(), **options)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/base.py", line 353, in send_task
    reply_to=reply_to or self.oid, **options
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/amqp.py", line 305, in publish_task
    **kwargs
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/messaging.py", line 172, in publish
    routing_key, mandatory, immediate, exchange, declare)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py", line 470, in _ensured
    interval_max)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py", line 382, in ensure_connection
    interval_start, interval_step, interval_max, callback)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/utils/__init__.py", line 246, in retry_over_time
    return fun(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py", line 250, in connect
    return self.connection
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py", line 756, in connection
    self._connection = self._establish_connection()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py", line 711, in _establish_connection
    conn = self.transport.establish_connection()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/transport/pyamqp.py", line 116, in establish_connection
    conn = self.Connection(**opts)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/connection.py", line 180, in __init__
    (10, 30),  # tune
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/abstract_channel.py", line 67, in wait
    self.channel_id, allowed_methods, timeout)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/connection.py", line 241, in _wait_method
    channel, method_sig, args, content = read_timeout(timeout)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/connection.py", line 330, in read_timeout
    return self.method_reader.read_method()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/method_framing.py", line 189, in read_method
    raise m
error: [Errno 104] Connection reset by peer
```

bdero commented:
@haikuginger Can you paste your variable output on OCIM?

haikuginger commented:
@bdero, will do so on our internal ticket, as the variable payload carries some secrets.

bdero commented:
This is ready for review again @haikuginger

haikuginger commented:
I am absolutely DELIGHTED to cast a conditional :+1: once integration tests are passing.

:+1:

- [x] I tested this: I was able to deploy an instance using a remote RabbitMQ host, as described, and both XQueue and Celery users showed activity on the vhost that was created for this instance. I was able to create a re-run of a course.
- [x] I read through the code
- [x] ~~I checked for accessibility issues~~ N/A; no UI changes
- [x] Includes documentation

bdero commented:
:tada:

pomegranited commented:
Hey @bdero, I'm having trouble running the unit tests on my devstack since this PR was merged.  I ended up having to add the `RABBITMQ_*` env settings to my .env.test file to get these three tests to pass:

```
RABBITMQ_API_URL='https://rabbitmq-dev.opencraft.hosting:15671'
RABBITMQ_ADMIN_USERNAME='admin'
RABBITMQ_ADMIN_PASSWORD='real password here'
```

Since the settings are in CircleCI, the tests pass there.  But since we can't put our real RABBITMQ credentials in our .env.test file for this repo, the actual service provisioning calls need to be mocked for the unit tests.  It's ok to keep them real for the integration tests though, because the integration tests require real services and credentials.

```
======================================================================
ERROR: test_rabbitmq_request (instance.tests.models.test_openedx_database_mixins.RabbitMQInstanceTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/opencraft/instance/tests/models/test_openedx_database_mixins.py", line 539, in test_rabbitmq_request
    response = self.instance._rabbitmq_request('get', 'overview')
  File "/home/vagrant/opencraft/instance/models/mixins/rabbitmq.py", line 90, in _rabbitmq_request
    raise RabbitMQAPIError
instance.models.mixins.rabbitmq.RabbitMQAPIError

======================================================================
FAIL: test_rabbitmq_access (instance.tests.models.test_openedx_database_mixins.RabbitMQInstanceTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/opencraft/instance/tests/models/test_openedx_database_mixins.py", line 531, in test_rabbitmq_access
    self.assertTrue(response.ok)
AssertionError: False is not true

======================================================================
ERROR: test_spawn_appserver_with_external_databases (instance.tests.models.test_openedx_instance.OpenEdXInstanceTestCase)
Wrap the test with appropriate mocks
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/opencraft/instance/tests/utils.py", line 145, in wrapper
    return func(self, mocks, *args, **kwargs)
  File "/usr/lib/python3.5/unittest/mock.py", line 1157, in patched
    return func(*args, **keywargs)
  File "/home/vagrant/opencraft/instance/tests/models/test_openedx_instance.py", line 415, in test_spawn_appserver_with_external_databases
    appserver_id = instance.spawn_appserver()
  File "/home/vagrant/opencraft/instance/logging.py", line 51, in wrapper
    return method(self, *args, **kwds)
  File "/home/vagrant/opencraft/instance/models/openedx_instance.py", line 362, in spawn_appserver
    self.provision_rabbitmq()
  File "/home/vagrant/opencraft/instance/models/mixins/rabbitmq.py", line 98, in provision_rabbitmq
    self._rabbitmq_request('put', 'vhosts', self.rabbitmq_vhost)
  File "/home/vagrant/opencraft/instance/models/mixins/rabbitmq.py", line 90, in _rabbitmq_request
    raise RabbitMQAPIError
instance.models.mixins.rabbitmq.RabbitMQAPIError

----------------------------------------------------------------------
```

The place to mock the service would be in the [`@patch_services` decorator](https://github.com/open-craft/opencraft/blob/master/instance/tests/utils.py#L80), and then apply that decorator to the above tests.

antoviaque commented:
@bdero When you look at this, could you also add the RabbitMQ settings to the dev `.env` file from our documentation repository?
